sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +334,212 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+      public void processElement(ProcessContext c) throws Exception {
+        ByteBuffer data = ByteBuffer.wrap(c.element());
+
+        String partitionKey = spec.getPartitionKey();
+        String explicitHashKey = null;
+
+        // Use custom partitioner
+        if (partitioner != null) {
+          partitionKey = partitioner.getPartitionKey(c.element());
+          explicitHashKey = partitioner.getExplicitHashKey(c.element());
         }
-        return clientBuilder.build();
+
+        ListenableFuture<UserRecordResult> f =
+            producer.addUserRecord(spec.getStreamName(), partitionKey, explicitHashKey, data);
+        Futures.addCallback(f, new UserRecordResultFutureCallback());
+      }
+
+      @FinishBundle
+      public void finishBundle() throws Exception {
+        // Flush all outstanding records, blocking call
+        flushSync();
+
+        // Check if outstanding records still exist
+        checkOutstandingRecords();
       }
 
-      @Override
-      public AmazonCloudWatch getCloudWatchClient() {
-        AmazonCloudWatchClientBuilder clientBuilder =
-            AmazonCloudWatchClientBuilder.standard().withCredentials(getCredentialsProvider());
-        if (serviceEndpoint == null) {
-          clientBuilder.withRegion(region);
-        } else {
-          clientBuilder.withEndpointConfiguration(
-              new AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region.getName()));
+      @Teardown
+      public void tearDown() throws Exception {
+        if (producer != null) {
+          producer.destroy();
+        }
+      }
+
+      /**
+       * Own implementation of {@link KinesisProducer#flushSync()} to prevent infinite loop in case
+       * of failures.
+       */
+      private void flushSync() {
+        int retries = spec.getRetries();
+
+        while (producer.getOutstandingRecordsCount() > 0 && retries-- > 0) {
+          producer.flush();
+          try {
+            Thread.sleep(1000);
+          } catch (InterruptedException e) {
+            LOG.warn("An exception occurred while flush: " + e);
+          }
+        }
+      }
+
+      /**
+       * Check if some non-flushed records records still exist.
+       *
+       * @throws IOException
+       */
+      private void checkOutstandingRecords() throws IOException {
+        int outstandingRecordsCount = producer.getOutstandingRecordsCount();
+        if (outstandingRecordsCount > 0) {
+          String message =
+              String.format(
+                  "Number of outstanding records are greater than zero: %d",
+                  outstandingRecordsCount);
+          LOG.error(message);
+          throw new IOException(message);
+        }
+      }
+
+      private static class UserRecordResultFutureCallback
+          implements FutureCallback<UserRecordResult> {
+
+        @Override
+        public void onFailure(Throwable t) {
+          LOG.error("An exception occurred while processing a record", t);
```

Comment:
Users don't look at error logs unless their pipeline is failing - errors must lead to exceptions, and a pipeline must not succeed unless it has successfully performed all the requested processing.

See https://beam.apache.org/contribute/ptransform-style-guide/#error-handling - specifically the part with "Never, ever, ever do this" :)

Please make sure that flushing throws these errors. See implementation of BigtableIO.write() for a similar pattern.

Reply:
+1 I left similar comment

Reply:
Yes, I did what I should never have to do =) Thank you for pointing out this. Actually, I think we need to do the same (throw an exception) in `onSuccess()` too in case if `result.isSuccessful() == false`. Are you agree with this?

Reply:
yes, throw an unchecked exception

Reply:
@echauchot @jkff I added a logic to throw an exception in case of failed records (it's based on BigtableIO code).